# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JustAnn1234/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

* **Selected Lane:** Lane 2 — Refresh / Content Opportunity Scoring
* **Why this lane:**
  Websites maintain hundreds or thousands of published content items. Over time, articles become outdated, lose rank to competitors, or see traffic decay. Manually auditing an entire content catalog to decide which pages deserve an update is inefficient and unscalable.

  This lane provides an actionable, decision-support workflow: using historical search performance and engagement metrics to build a ranked review queue for content updates. Additionally, by incorporating observable AI referral signals (`ai_sessions_90d` and `ai_traffic_pct`) alongside traditional search metrics, we can evaluate whether pages losing traditional search visibility retain value across modern AI engines.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

### Search Question
> *"Which published web pages exhibit significant performance decay or untapped traffic demand, and should be prioritized first by the content team for a refresh or rewrite?"*

### Core Framing Dimensions
* **Unit of Analysis (Grain):** A single unique content item (`content_id` / `content_hash_id`) aggregated over a 90-day feature window.
* **Model Output:** A continuous **Priority Refresh Score** (0–100) and probability of decay, paired with human-readable reason codes (e.g., `declining_with_demand`, `stale_visible_page`).
* **Decision:** Deciding which top 5%–10% of declining or high-demand content items should be assigned to editorial/growth teams for updates next month.
* **Action:** The content team reviews the flagged pages, updates outdated facts, expands thin sections, or restructures headings/meta descriptions.
* **Cost of a Wrong Call:**
  * **False Positive (High priority assigned to a non-performing or irrecoverable page):** Wasted editorial budget and reviewer time spent updating a page that yields zero traffic recovery.
  * **False Negative (High-potential declining page missed):** Organic search traffic continues to decay, causing a persistent loss of site traffic to competitors.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import pandas as pd
import numpy as np

# List of relative local paths to try, followed by a raw GitHub fallback
data_url = "https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv"

paths_to_try = [
    '../data/raw/content_refresh_anonymized.csv',
    'data/raw/content_refresh_anonymized.csv',
    '../../data/raw/content_refresh_anonymized.csv',
    data_url
]

df = None
for path in paths_to_try:
    try:
        df = pd.read_csv(path)
        print(f"Successfully loaded data from: {path}")
        break
    except Exception:
        continue

if df is None:
    raise FileNotFoundError("Could not load content_refresh_anonymized.csv from local paths or GitHub.")

# --- Number 1: Total Scale & Scope ---
total_rows = len(df)

# --- Number 2: Opportunity Volume (Declining with Demand) ---
declining_pages = df[(df['impressions_90d'] >= 100) & (df['trend_direction'] == 'down')]
pct_declining = (len(declining_pages) / total_rows) * 100

# --- Number 3: AI Traffic Footprint ---
ai_active_pages = df[df['ai_sessions_90d'] > 0]
pct_ai_active = (len(ai_active_pages) / total_rows) * 100

print("\n=== STARTER DATASET SUMMARY METRICS ===")
print(f"1. Total Content Items Analyzed: {total_rows:,} rows")
print(f"2. Refresh Candidates (Impressions >= 100 & Trend == 'down'): {len(declining_pages):,} pages ({pct_declining:.1f}%)")
print(f"3. AI Traffic Exposure (Pages with >0 AI Sessions): {len(ai_active_pages):,} pages ({pct_ai_active:.1f}%)")

Successfully loaded data from: https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv

=== STARTER DATASET SUMMARY METRICS ===
1. Total Content Items Analyzed: 30,000 rows
2. Refresh Candidates (Impressions >= 100 & Trend == 'down'): 13,152 pages (43.8%)
3. AI Traffic Exposure (Pages with >0 AI Sessions): 1,930 pages (6.4%)


### Why these numbers justify the lane:
1. **Massive Opportunity Volume:** Exactly **13,152 pages (43.8%)** in the dataset maintain active search demand (`impressions_90d >= 100`) yet show a downward trend (`trend_direction == 'down'`). Manually auditing over 13,000 declining pages is unfeasible for a content team, proving the clear need for an automated priority ranking system.
2. **AI Traffic Analysis Integration:** A total of **1,930 pages (6.4%)** display active AI referral sessions (`ai_sessions_90d > 0`). This provides a strong sample size to analyze how traditional search decay interacts with AI engine traffic.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

##  Careful Words: What I Can and Can't Claim

* **What I CAN claim:**
  * The model ranks content items based on historical visibility loss, demand levels, and observable performance signals.
  * The resulting list provides a prioritized decision-support queue for human reviewers.
  * The metrics identify observational associations between page attributes (age, word count, CTR) and traffic trends.

* **What I CANNOT claim:**
  * **Causality:** I cannot guarantee that updating a page will cause Google to restore its search rankings or traffic.
  * **Search Engine Algorithm Factors:** I cannot claim to have discovered underlying ranking factors in search engine algorithms.
  * **Guaranteed AI Citations:** AI referral traffic counts only reflect recorded user click-throughs from AI tools, not total underlying AI citations or brand mentions.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

| Self-Check Criteria | Status | Notes |
| :--- | :---: | :--- |
| Picked a predefined lane or declared freestyle | **YES** | Selected Refresh / Content Opportunity Scoring |
| Defined unit of analysis, output, decision, and action | **YES** | Page grain, priority score, review decision, content update action |
| Highlighted the cost of a wrong call | **YES** | Addressed false positive (wasted editor time) & false negative (lost traffic) |
| Included 2–3 real numbers from the starter dataset | **YES** | Printed total rows, declining page %, and AI session footprint |
| Uses disciplined, honest language | **YES** | Clarified decision-support nature vs. causal guarantees |

#### Submission Checklist Confirmation:
- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/w01_research_question.ipynb`